### DISCUSSION TITLE
MDX Improvements

### DISCUSSION THREAD

**Discussion by Ehesp** [vercel/next.js#11043] [Help]
Hey,

I've built (and building) some fairly large static documentation websites which have both manual and auto generated API reference content. Right now I'm using Gatsby, which involves GraphQL - it's a pain to work with due to the complex nature of the auto-generated content & poses some limitations.

I believe the latest Next update could get around those issues, however from my research, I don't believe the MDX aspect to quite powerful enough:

---

1)

I use `gatsby-plugin-mdx` which allows me to convert the Markdown, however unlike NextJS, I can manually inject the content into a component using the `MDXProvider`, e.g:

```js
<MDXProvider components={components}>
  <MDXRenderer>{mdx.body}</MDXRenderer>
</MDXProvider>
```

`mdx.body` is the MDX parsed string from the plugin, and `components` is my custom object of global components which the MDX uses to render into HTML.

Is there a way of having such functionality with NextJS? I'd like to take the content of the MDX output and use it inside of a React component like above, rather than Next automatically generate pages.

I'm personally not a fan of putting layout inside of my MDX, hence the need to pass it as a prop to a custom component.

---

2) 

I take advantage of frontmatter quite a lot, however currently it seems this is lost. I realise I could export an object, however to me this doesn't feel "markdown" like, especially considering the content can be edited by the public.

Using Gatsby, I can access the frontmatter as an object during the creation of the pages.

I've seen https://github.com/hashicorp/next-mdx-enhanced, which I need to look into more, however I'm not sure whether it addresses point 1 still.

Cheers

**Comment by kvnwolf** [comment]
I guess you can just remove the mdx plugin and instead add a raw loader for `.mdx` files, and put your mdx files on a folder like `content/posts`.

Then, you can create an utility function to parse an mdx file and extract the frontmatter out of it with https://www.npmjs.com/package/gray-matter. That function can return an object containing `frontMatter` and `body` properties, and parse the `body` property with [`@mdx-js/runtime`](https://www.npmjs.com/package/@mdx-js/runtime).

So, in resume:

1. Create a file within `pages/posts/index.js`, which exports a `getStaticProps` function that reads the `content/posts` directory contents with `require.context`, then maps the array and run `parseMdx` on each file, then passes the posts to the page component, in order to list posts.

2. Create a file within `pages/posts/[slug].js`, which exports a `getStaticPaths` function that also reads the files of `content/posts` and creates paths for each blogpost. Now use the mdx runtime on the page component to transform that markdown to HTML.

**Reply by Ehesp** [reply]
Interesting approach, I didn't really think of doing it that way. I guess I'll need to dig into which plugins are used.

**Comment by timedtext** [comment]
1. Here is a example  about Custom Layout [with-app-layout](https://github.com/zeit/next.js/tree/canary/examples/with-app-layout)

2. There also a example about Front Matter on MDX's official website [mdx-custom-loader](https://mdxjs.com/guides/custom-loader)

**Comment by ldalves** [comment]
hey @Ehesp! 👋  i know this has been a while, but I wonder if you found a solution for this issue?
